examples_tests/06.MeshLoaders/main.cpp

```diff
@@ -7,81 +7,188 @@
+#define NABLA_TRANSFER_QUEUE_INDEX 0    // index within family
+#define NABLA_COMPUTE_QUEUE_INDEX 0     // index within family
+#define NABLA_SHARING_MODE ::nbl::asset::ESM_EXCLUSIVE
+#define NABLA_SWAPCHAIN_IMAGE_FORMAT ::nbl::asset::EF_R8G8B8A8_SRGB
+
+inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::video::E_DEBUG_MESSAGE_TYPE type, const char* msg, void* userData)
+{
+    const char* sev = nullptr;
+
+    switch (severity)
+    {
+        case video::EDMS_VERBOSE:
+        {
+            sev = "verbose";
+        } break;
+
+        case video::EDMS_INFO:
+        {
+            sev = "info";
+        } break;
+
+        case video::EDMS_WARNING:
+        {
+            sev = "warning";
+        } break;
+
+        case video::EDMS_ERROR:
+        {
+            sev = "error";
+        } break;
+    }
+
+    std::cout << "OpenGL " << sev << ": " << msg << std::endl;
+}
+
 int main()
 {
-	// create device with full flexibility over creation parameters
-	// you can add more parameters if desired, check nbl::SIrrlichtCreationParameters
-	nbl::SIrrlichtCreationParameters params;
-	params.Bits = 24; //may have to set to 32bit for some platforms
-	params.ZBufferBits = 24; //we'd like 32bit here
-	params.DriverType = video::EDT_OPENGL; //! Only Well functioning driver, software renderer left for sake of 2D image drawing
-	params.WindowSize = dimension2d<uint32_t>(1280, 720);
-	params.Fullscreen = false;
-	params.Vsync = true; //! If supported by target platform
-	params.Doublebuffer = true;
-	params.Stencilbuffer = false; //! This will not even be a choice soon
-	auto device = createDeviceEx(params);
-
-	if (!device)
+    constexpr uint32_t WIN_W = 800u;
+    constexpr uint32_t WIN_H = 600u;
+    constexpr uint32_t SC_IMG_COUNT = 3u;
+
+    auto window = CWindowT::create(WIN_W, WIN_H, ui::IWindow::ECF_NONE);
+
+    video::SDebugCallback debugCallbackObject;
+    debugCallbackObject.callback = &debugCallback;
+    debugCallbackObject.userData = nullptr;
+
+    auto glApiConnection = video::IAPIConnection::create(video::EAT_OPENGL, 0, "New API Test", debugCallbackObject);
+    auto surface = glApiConnection->createSurface(window.get());
+
+    auto physicalDevices = glApiConnection->getPhysicalDevices();
+    assert(!physicalDevices.empty());
+
+    nbl::core::smart_refctd_dynamic_array<uint32_t> queueFamilyIxs = nullptr;
+    uint32_t queueFamilyIx_render = 0xffffFFFFu;
+    uint32_t queueFamilyIx_compute = 0xffffFFFFu;
+    uint32_t queueFamilyIx_transfer = 0xffffFFFFu;
+
+    auto gpuPhysicalDevice = physicalDevices.begin()[0];
+    {
+        auto queueFamilyPropertiesRange = gpuPhysicalDevice->getQueueFamilyProperties();
+        for (uint32_t i = 0u; i < queueFamilyPropertiesRange.size(); ++i)
+        {
+            const auto& queueFamilyProperties = queueFamilyPropertiesRange.begin()[i];
+
+            if (queueFamilyProperties.queueFlags & nbl::video::IPhysicalDevice::EQF_GRAPHICS_BIT)
```

Comment:
could I get vulkan docs reference to this one?

Reply:
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkGetPhysicalDeviceQueueFamilyProperties.html

Reply:
all this set up and device choice should be hidden inside `CommonAPI.h` that @sadiuk made

Reply:
of course

Reply:
@devshgraphicsprogramming I'm not a vulkan spec, but I'm pretty sure I can't use this utilities, look at what my code does with families while filling `queuesParams` https://github.com/Devsh-Graphics-Programming/Nabla/blob/284ef538d096d3a439d8c74fa5e8132935b9b275/examples_tests/06.MeshLoaders/main.cpp#L100. First of all I assume 3 families (needed) and the common API invoke bit alternative for getting family index with 3 bits for either graphics, compute and transfer, **but a assigns only 1 family anyways in queueParamsCount**

Reply:
look at CommonAPI https://github.com/Devsh-Graphics-Programming/Nabla/blob/284ef538d096d3a439d8c74fa5e8132935b9b275/examples_tests/common/CommonAPI.h#L55

Reply:
it seems like CommonAPI supported only `video::IPhysicalDevice::EQF_GRAPHICS_BIT`

Reply:
write method for `CommonAPI` to give you queues for:
- Graphics (if commonAPI initialized with boolean saying graphics work will be performed)
- Compute
- Transfer Up
- Transfer Down

Right now create all from the same family, and only create one queue (return the same queue for all `get`s)

Using separate queues (and even families) is not always a win, and it makes almost 0 sense on OpenGL (except for transfers) so we'll wait until Vulkan to change what the default stuff on examples is.

Reply:
@AnastaZIuk done?

Reply:
done in following commit